libavformat/rtcenc.c

```diff
@@ -0,0 +1,1548 @@
+    if (st->codecpar->codec_type == AVMEDIA_TYPE_AUDIO) {
+        pkt->dts = pkt->pts = rtc->audio_jitter_base;
+        rtc->audio_jitter_base += 960;
+    }
+
+    /* Insert a packet with SPS/PPS before each IDR frame. */
+    is_idr = (pkt->flags & AV_PKT_FLAG_KEY) && st->codecpar->codec_type == AVMEDIA_TYPE_VIDEO;
+    if (is_idr && st->codecpar->extradata) {
+        extra = av_packet_alloc();
+        size = !rtc->avc_nal_length_size ? st->codecpar->extradata_size :
+                rtc->avc_nal_length_size * 2 + rtc->avc_sps_size + rtc->avc_pps_size;
+        ret = av_new_packet(extra, size);
+        if (ret < 0) {
+            av_log(s, AV_LOG_ERROR, "Failed to allocate extra packet\n");
+            return ret;
+        }
+
+        /* Encode SPS/PPS in annexb format. */
+        if (!rtc->avc_nal_length_size) {
+            memcpy(extra->data, st->codecpar->extradata, size);
+        } else {
+            /* Encode SPS/PPS in ISOM format. */
+            p = extra->data;
+            for (i = 0; i < rtc->avc_nal_length_size; i++) {
+                *p++ = rtc->avc_sps_size >> (8 * (rtc->avc_nal_length_size - i - 1));
+            }
+            memcpy(p, rtc->avc_sps, rtc->avc_sps_size);
+            p += rtc->avc_sps_size;
+
+            /* Encode PPS in ISOM format. */
+            for (i = 0; i < rtc->avc_nal_length_size; i++) {
+                *p++ = rtc->avc_pps_size >> (8 * (rtc->avc_nal_length_size - i - 1));
+            }
+            memcpy(p, rtc->avc_pps, rtc->avc_pps_size);
+            p += rtc->avc_pps_size;
+        }
+
+        /* Setup packet and feed it to chain. */
+        extra->pts = pkt->pts;
+        extra->dts = pkt->dts;
+        extra->stream_index = pkt->stream_index;
+        extra->time_base = pkt->time_base;
+
+        ret = ff_write_chained(rtp_ctx, 0, extra, s, 0);
+        if (ret < 0)
+            goto end;
+    }
+
+    ret = ff_write_chained(rtp_ctx, 0, pkt, s, 0);
+
+end:
+    av_packet_free(&extra);
+    return ret;
+}
+
+static int rtc_write_trailer(AVFormatContext *s)
+{
+    return 0;
+}
+
+static av_cold void rtc_deinit(AVFormatContext *s)
+{
+    int i;
+    RTCContext *rtc = s->priv_data;
+
+    for (i = 0; i < s->nb_streams; i++) {
+        AVFormatContext* rtp_ctx = s->streams[i]->priv_data;
+        if (!rtp_ctx)
+            continue;
+
+        av_write_trailer(rtp_ctx);
+        avio_context_free(&rtp_ctx->pb);
+        avformat_free_context(rtp_ctx);
+        s->streams[i]->priv_data = NULL;
+    }
+
+    av_freep(&rtc->avc_sps);
+    av_freep(&rtc->avc_pps);
+    av_freep(&rtc->sdp_offer);
+    av_freep(&rtc->sdp_answer);
+    av_freep(&rtc->ice_ufrag_remote);
+    av_freep(&rtc->ice_pwd_remote);
+    av_freep(&rtc->ice_protocol);
+    av_freep(&rtc->ice_host);
+    ffurl_closep(&rtc->udp_uc);
+    ff_srtp_free(&rtc->srtp_audio_send);
+    ff_srtp_free(&rtc->srtp_video_send);
+    ff_srtp_free(&rtc->srtp_recv);
+}
+
+static const AVOption options[] = {
+    { NULL },
+};
```

Comment:
`/* The maximum size of an SDP, either offer or answer. */
#define MAX_SDP_SIZE 8192
/* The maximum size of a UDP packet, should be smaller than the MTU. */
#define MAX_UDP_SIZE 1500
/**
 * The RTP payload max size, reserved some paddings for SRTP as such:
 *      kRtpPacketSize = kRtpMaxPayloadSize + paddings
 * For example, if kRtpPacketSize is 1500, recommend to set kRtpMaxPayloadSize to 1400,
 * which reserves 100 bytes for SRTP or paddings.
 * otherwise, the kRtpPacketSize must less than MTU, in webrtc source code,
 * the rtp max size is assigned by kVideoMtu = 1200.
 * so we set kRtpMaxPayloadSize = 1200.
 * see @doc https://groups.google.com/g/discuss-webrtc/c/gH5ysR3SoZI
 */
#define MAX_UDP_PAYLOAD_SIZE (MAX_UDP_SIZE - 300)
/* The maximum number of retries for UDP transmission. */
#define UDP_FAST_RETRIES 6
/* The startup timeout for UDP transmission. */
#define UDP_START_TIMEOUT 21
/* Avoid dtls negotiate failed, set max fragment size 1200. */
#define DTLS_FRAGMENT_MAX_SIZE 1200
/* Supported DTLS cipher suites. */
#define DTLS_CIPHER_SUTES "ECDHE-ECDSA-AES128-GCM-SHA256"\
    ":ECDHE-RSA-AES128-GCM-SHA256"\
    ":ECDHE-ECDSA-AES128-SHA"\
    ":ECDHE-RSA-AES128-SHA"\
    ":ECDHE-ECDSA-AES256-SHA"\
    ":ECDHE-RSA-AES256-SHA"
/* The SRTP key size, defined by SRTP_MASTER_KEY_LEN */
#define DTLS_SRTP_MASTER_KEY_LEN 30
/* The NALU type for STAP-A */
#define NALU_TYPE_STAP_A 24`

can these controlled by options?

Reply:
If the value may change in different use scenario, it should be an option. For a constant value that never changes, it should be defined by macro, rather than an option.

The bellow values literally never change or user doesn't know how to set it:

1. `#define MAX_SDP_SIZE 8192`: It never changes and 8k is big enough. Rather than an option, it should be a larger value if not enough.
2. `#define DTLS_CIPHER_SUTES ECDHE-ECDSA-AES128-GCM-SHA256`: Depends on the ability of ssl library, user doesn't know how to set it.
3. `#define DTLS_SRTP_MASTER_KEY_LEN 30`: Never changes.
4. `#define NALU_TYPE_STAP_A 24`: Never changes.

The bellow values should be FFmpeg options because it may change:

1. `#define MAX_UDP_SIZE 1500`: Generally, the network MTU is 1500, but it may be smaller for some special and legacy network, so it should be an option, rather than a macro.
2. `#define UDP_FAST_RETRIES 6`: Should be an option because user may change it to retry more times.
5. `#define UDP_START_TIMEOUT 21`: Should be an option because user may change it to use larger timeout.
6. `#define DTLS_FRAGMENT_MAX_SIZE 1200`: Should be an option for it's similar to `MAX_UDP_SIZE`.